Fix user methods requiring a session in Panache REST Data with Reactive

Sgitario opened the pull request:
These changes add the `@WithSessionOnDemand` annotation at class level when using Panache REST Data with Hibernate Reactive. 

This annotation makes user methods that return an Uni class to work.

About tests, I've covered the following scenarios:
- user GET method that returns an Uni
- user POST method that returns an Uni (and requires a transaction)
- injecting the generated resource will also work

Fix https://github.com/quarkusio/quarkus/issues/34432

geoand reviewed (approved):
LGTM

mkouba commented on extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/rest/data/panache/deployment/entity/CollectionsResource.java:
```diff
@@ -1,8 +1,32 @@
+import io.smallrye.mutiny.Uni;
 
 @ResourceProperties(hal = true, paged = false, halCollectionName = "item-collections")
+@WithSessionOnDemand
```
This annotation should not be necessary here, or?

Sgitario replied:
This was actually a leftover, thanks for spotting!

Sgitario replied:
PR updated

mkouba commented on extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/JaxRsResourceImplementor.java:
```diff
@@ -107,6 +108,9 @@ private void implementClassAnnotations(ClassCreator classCreator, ResourceMetada
                 classCreator.addAnnotation(classAnnotation);
             }
         }
+        if (capabilities.isPresent(Capability.HIBERNATE_REACTIVE)) {
```
I don't know how Panache REST Data works internally, but why is this actually needed? Is this class reused by the reactive version as well?

Sgitario replied:
Exactly, this implementor is used also for HIBERNATE classic, so we need to check the capabilities.

mkouba commented on extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/ParamData.java:
```diff
@@ -0,0 +1,27 @@
+ * If we're filling parameters for a sub-client, we need to know the defining class of the ClientxxxParam
+ * to properly resolve default methods of the "root" client
+ */
+public class ParamData {
```
Where's this class used?

Sgitario replied:
Sorry again, this was part of another work I was doing and added here by mistake. Removed.

mkouba replied:
No problem :-)

Sgitario committed:
Fix user methods requiring a session in Panache REST Data with Reactive

These changes add the `@WithSessionOnDemand` annotation at class level when using Panache REST Data with Hibernate Reactive. 

This annotation makes user methods that return an Uni class to work.

About tests, I've covered the following scenarios:
- user GET method that returns an Uni
- user POST method that returns an Uni (and requires a transaction)
- injecting the generated resource will also work

Fix https://github.com/quarkusio/quarkus/issues/34432

gsmet commented:
Do we want this backported to 3.2? If so please add the backport label.

Sgitario commented:
> Do we want this backported to 3.2? If so please add the backport label.

I would not backport this since there is an existing workaround.

Dieken commented:
Isn’t 3.2 a LTS version?  I would like it’s backported, the explicit call to `Panache.withSession` is very ugly, does’t match the non-reactive version.

Dieken commented:
@Sgitario   could you consider again to backport this fix to 3.2?   If 3.2 will be a LTS version,  the workaround is very ugly, I would like to have consistent developer experience with non-reactive version.

Sgitario commented:
> @Sgitario could you consider again to backport this fix to 3.2? If 3.2 will be a LTS version, the workaround is very ugly, I would like to have consistent developer experience with non-reactive version.

These changes introduce a very major change from a technical view perspective, so I would like to iterate over the solution a bit more, to make sure everything works as before.

Sgitario commented:
@Dieken since this solution seems to work fine in some test iterations (for a few days). I've just added the `backport` label to incorporate these changes into 3.2.
